python_modules/dagster/dagster/_core/definitions/freshness_policy_sensor_definition.py

```diff
@@ -0,0 +1,366 @@
+                    )
+
+                if result is not None:
+                    raise DagsterInvalidDefinitionError(
+                        "Functions decorated by `@freshness_policy_sensor` may not return or yield a value."
+                    )
+
+            context.update_cursor(
+                FreshnessPolicySensorCursor.from_dict(minutes_late_by_key).to_json()
+            )
+
+        super(FreshnessPolicySensorDefinition, self).__init__(
+            name=name,
+            evaluation_fn=_wrapped_fn,
+            minimum_interval_seconds=minimum_interval_seconds,
+            description=description,
+            default_status=default_status,
+        )
+
+    def __call__(self, *args, **kwargs):
+        if is_context_provided(self._freshness_policy_sensor_fn):
+            if len(args) + len(kwargs) == 0:
+                raise DagsterInvalidInvocationError(
+                    "Freshness policy sensor function expected context argument, but no context argument "
+                    "was provided when invoking."
+                )
+            if len(args) + len(kwargs) > 1:
+                raise DagsterInvalidInvocationError(
+                    "Freshness policy sensor invocation received multiple arguments. Only a first "
+                    "positional context parameter should be provided when invoking."
+                )
+
+            context_param_name = get_function_params(self._freshness_policy_sensor_fn)[0].name
+
+            if args:
+                context = check.opt_inst_param(
+                    args[0], context_param_name, FreshnessPolicySensorContext
+                )
+            else:
+                if context_param_name not in kwargs:
+                    raise DagsterInvalidInvocationError(
+                        f"Freshness policy sensor invocation expected argument '{context_param_name}'."
+                    )
+                context = check.opt_inst_param(
+                    kwargs[context_param_name], context_param_name, FreshnessPolicySensorContext
+                )
+
+            if not context:
+                raise DagsterInvalidInvocationError(
+                    "Context must be provided for direct invocation of freshness policy sensor."
+                )
+
+            return self._freshness_policy_sensor_fn(context)
+
+        else:
+            raise DagsterInvalidDefinitionError(
+                "Freshness policy sensor must accept a context argument."
+            )
+
+
+def freshness_policy_sensor(
+    asset_selection: AssetSelection,
+    *,
+    name: Optional[str] = None,
+    minimum_interval_seconds: Optional[int] = None,
+    description: Optional[str] = None,
+    default_status: DefaultSensorStatus = DefaultSensorStatus.STOPPED,
+) -> Callable[[Callable[[FreshnessPolicySensorContext], None]], FreshnessPolicySensorDefinition,]:
+    """
+    Define a sensor that reacts to the status of a given set of asset freshness policies, where the
```

Comment:
I kind of expected this to only get called when something became late. Are there non-late situations where you imagine this would get used? What would a canonical implementation look like?

Reply:
I was imagining a pattern like "alert a single time once I start being late, then alert a single time once I'm back to ok".

Reply:
```
@freshness_policy_sensor(
    asset_selection=AssetSelection.all(),
)
def slack_on_freshness_policy_status(context: FreshnessPolicySensorContext):
    if context.minutes_late > 10 and context.previous_minutes_late < 10:
        # send failure slack message
        pass
    if context.minutes_late == 0 and context.previous_minutes_late != 0:
        # send success slack message
        pass
```